pynest/examples/hpc_benchmark.py

```diff
@@ -0,0 +1,426 @@
+    logger = Logger(params['log_file'])
+
+    nest.ResetKernel
+    nest.set_verbosity(M_INFO)
+
+    logger.log(str(memory_thisjob()) + ' # virt_mem_0')
+
+    build_network(logger)
+
+    tic = time.time()
+
+    nest.Simulate(params['presimtime'])
+
+    PreparationTime = time.time() - tic
+
+    logger.log(str(memory_thisjob()) + ' # virt_mem_after_presim')
+    logger.log(str(PreparationTime) + ' # presim_time')
+
+    tic = time.time()
+
+    nest.Simulate(params['simtime'])
+
+    SimCPUTime = time.time() - tic
+
+    logger.log(str(memory_thisjob()) + ' # virt_mem_after_sim')
+    logger.log(str(SimCPUTime) + ' # sim_time')
+
+    logger.log(str(compute_rate()) + ' # average rate')
+
+    logger.done()
+
+# ------------------------------------------------------------------------------------
+
+
+def compute_rate():
+    '''compute approximation of average firing rate based on number of
+    local nodes, number of local spikes and total time
+    '''
+
+    n_local_spikes = nest.GetKernelStatus('local_spike_counter')
+    n_local_neurons = nest.GetKernelStatus('network_size') / nest.GetKernelStatus('num_processes')
+    simtime = nest.GetKernelStatus('time')
+    return 1. * n_local_spikes / (n_local_neurons * simtime) * 1e3
+
+#  ------------------------------------------------------------------------------------
+
+
+def memory_thisjob():
+    '''wrapper to obtain current memory usage'''
+    nest.sr('memory_thisjob')
+    return nest.spp()
+
+#  ------------------------------------------------------------------------------------
+
+
+def lambertwm1(x):
+    nest.sr('{} LambertWm1'.format(x))
+    return nest.spp()
+
+#  ------------------------------------------------------------------------------------
+
+
+def get_local_nodes(nodes):
+
+    '''
+    Generator for efficient looping over local nodes.
+    '''
+
+    nvp = nest.GetKernelStatus('total_num_virtual_procs')  # step size
+
+    i = 0
+    while i < len(nodes):
+        if nest.GetStatus([nodes[i]], 'local')[0]:
+            yield nodes[i]
+            i += nvp
+        else:
+            i += 1
+
+#  ------------------------------------------------------------------------------------
+
+
+class Logger(object):
+    '''
+    This class defines a logger class used to properly log memory and timing
+    information from network simulations. It is used by hpc_benchmark.sli to
+    store the information to the log files.
+    '''
+
+    def __init__(self, file_name):
+        self.max_rank_cout = 5  # copy output to cout for ranks 0..max_rank_cout-1
+        self.max_rank_log = 30  # write to log files for ranks 0..max_rank_log-1
+        self.line_counter = 0
+
+        if nest.Rank() < self.max_rank_log:
+
+            # convert rank to string, prepend 0 if necessary to make
+            # numbers equally wide for all ranks
+            rank = '{:0' + str(len(str(self.max_rank_log))) + '}'
+            fn = file_name + '_' + rank.format(nest.Rank()) + '.dat'
```

Comment:
It would be better to use proper string formatting here. For instance:

`fn = "{}_{}.dat".format(file_name, rank.format(nest.Rank))`

The same for the `log`-Function below.